### PR TITLE
[SpaceEngine] `intersect` cleanups

### PR DESCRIPTION
**Pull request by xedin**
Includes new `intersect` method which always returns simplified space and extracted `decompose` method which returns disjunct from decomposition of the type.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

File: lib/Sema/TypeCheckSwitchStmt.cpp [hunk 1]
```diff
@@ -609,21 +594,21 @@ namespace {
...
           }
 
           return Space::forDisjunct(paramSpace);
         }
 
-        PAIRCASE (SpaceKind::UnknownCase, SpaceKind::Type):
-        PAIRCASE (SpaceKind::UnknownCase, SpaceKind::Constructor):
-          return other.intersect(*this, TC, DC);
+        PAIRCASE(SpaceKind::UnknownCase, SpaceKind::Type):
+        PAIRCASE(SpaceKind::UnknownCase, SpaceKind::Constructor):
```
Comment (CodaFi) [inline lib/Sema/TypeCheckSwitchStmt.cpp]: Nit: Could you put the spaces back?

**Review by CodaFi — APPROVED** [review]
Modulo formatting nit, LGTM

File: lib/Sema/TypeCheckSwitchStmt.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/TypeCheckSwitchStmt.cpp]: Ah, missed that one, thanks!

**Comment by xedin** [comment]
@swift-ci please test

**Pull request merged**